Fix event overlay regions for some websites

mwyrzykowski opened the pull request:
#### ccc3cf9168e4dfe44b314caeaa38867c3dde0eb7
<pre>
Fix event overlay regions for some websites
<a href="https://bugs.webkit.org/show_bug.cgi?id=242830">https://bugs.webkit.org/show_bug.cgi?id=242830</a>

Reviewed by Tim Horton.

Rectangle being passed was not in the web view&apos;s coordinate space.

Transform the rectangle into the correct coordinate space.

Also fix KVO notification where navigating between web pages would
not send a notification for overlayRegions. This requires storing
a copy of the rectangles in the scroll view representing the current
state to observe changes.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(addOverlayEventRegions):
Correctly map the rectangle into the web view&apos;s coordinate space.

(-[WKWebView _updateOverlayRegions:destroyedLayers:]):
Pass the web view to the recursive function.

We could use iterators instead of creating a temporary vector of
CGRect instances but this vector is generally very small, 5-8 objects
in the worser cases, so it probably is not necessary.

(-[WKWebView _overlayRegions]): Deleted.
Done by WKScrollView now.

* Source/WebKit/UIProcess/ios/WKScrollView.h:
* Source/WebKit/UIProcess/ios/WKScrollView.mm:
(-[WKScrollView _updateOverlayRegions:]):
Post KVO notification if the data changes.

(-[WKScrollView overlayRegions]):
Getter returning an array of NSData instances.

Canonical link: <a href="https://commits.webkit.org/252680@main">https://commits.webkit.org/252680@main</a>
</pre>

hortont424 commented on Source/WebKit/UIProcess/ios/WKScrollView.mm:
```diff
@@ -549,10 +552,31 @@ - (CGPoint)_puic_contentOffsetForCrownInputSequencerOffset:(double)sequencerOffs
+    return true;
+}
+
+- (const Vector<CGRect> &)_overlayRegions
```
I don't love having the two getters with the different types just differentiated by a leading underscore. Anything we can put in the name to differentiate?

hortont424 replied:
(or, better, can we just make the NSArray when they're set and only have one getter?)

mwyrzykowski replied:
Oh I don't even use the `_overlayRegions` version, I should just remove that

hortont424 commented on Source/WebKit/UIProcess/ios/WKScrollView.mm:
```diff
@@ -549,10 +552,31 @@ - (CGPoint)_puic_contentOffsetForCrownInputSequencerOffset:(double)sequencerOffs
 - (NSArray<NSData *> *)overlayRegions
 {
-    return [_internalDelegate _overlayRegions];
+    NSMutableArray<NSData *> *overlayRegions = [NSMutableArray array];
```
I think this might be more efficient if it used VectorCocoa's `createNSArray(CollectionType&& collection, MapFunctionType&& function)` (if it can)

mwyrzykowski replied:
```
    return createNSArray(_overlayRegions, [] (auto& rect) -> NSData * {
        return [NSData dataWithBytes:(void*)&rect length:sizeof(rect)];
    });
```
that should work

mwyrzykowski committed:
Fix event overlay regions for some websites
https://bugs.webkit.org/show_bug.cgi?id=242830

Reviewed by Tim Horton.

Rectangle being passed was not in the web view's coordinate space.

Transform the rectangle into the correct coordinate space.

Also fix KVO notification where navigating between web pages would
not send a notification for overlayRegions. This requires storing
a copy of the rectangles in the scroll view representing the current
state to observe changes.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(addOverlayEventRegions):
Correctly map the rectangle into the web view's coordinate space.

(-[WKWebView _updateOverlayRegions:destroyedLayers:]):
Pass the web view to the recursive function.

We could use iterators instead of creating a temporary vector of
CGRect instances but this vector is generally very small, 5-8 objects
in the worser cases, so it probably is not necessary.

(-[WKWebView _overlayRegions]): Deleted.
Done by WKScrollView now.

* Source/WebKit/UIProcess/ios/WKScrollView.h:
* Source/WebKit/UIProcess/ios/WKScrollView.mm:
(-[WKScrollView _updateOverlayRegions:]):
Post KVO notification if the data changes.

(-[WKScrollView overlayRegions]):
Getter returning an array of NSData instances.

Canonical link: https://commits.webkit.org/252680@main
